crypto/crypto.lurk

```diff
@@ -0,0 +1,96 @@
+; crypto.lurk
+; 
+; simple crypto-related functions
+;
+; load in repl as:
+; > :load "crypto.lurk"
+
+
+(letrec (
+
+    ; parity tests
+    (odd (lambda (a) (< (/ a 2) 0)))
+    (even (lambda (a) (eq NIL (odd a))))
+
+    ; square & multiply exponentiation
+    (fastexp (lambda (b e)
+        (if (= e 0) 1
+            (if (even e) (fastexp (* b b) (/ e 2))
+                (* b (fastexp (* b b) (/ (- e 1) 2)))))))
+
+    ; GCD
+    (gcd (lambda (a b) 
+        (if (= a b) a
+            (if (> a b) (gcd (- a b) b)
+                (gcd (- b a) a)))))
+
+    ; hash function, poseidon-like
+    ; 2-element permutation used by the hash
+    ; NOT cryptographically safe
+    (permute (lambda (a b)
+        (letrec ( 
+            (rnd (lambda (i a b) (let (
+                (mat1 (lambda (a b) (+ (/ a 3) (* b 5))))
+                (mat2 (lambda (a b) (+ (/ a 7) (* b 9))))
+                (sbox (lambda (a) (fastexp a 5)))
+                (add1 (lambda (a) (+ a (/ 2 3))))
+                (add2 (lambda (a) (+ a (/ 3 2))))
+                )
+                (cons
+                    (+ i (add1 (sbox (mat1 a b))))
+                    (add2 (sbox (mat2 a b)))))))
+            (iter (lambda (i a b)
+                (if (= i 4) ; 4 rounds
+                    (cons a b)
+                    (let (
+                        (res (rnd i a b))
+                        (a (car res))
+                        (b (cdr res))
+                        )
+                        (iter (+ i 1) a b)))))) 
+            (iter 0 a b))))
+
+    ; initial state of the hash will be (0 iv)
+    (iv (/ 5 2))
+
+    ; hash 1 element
+    (hash1 (lambda (a) (car (permute a iv))))
+
+    ; hash 2 elements
+    (hash2 (lambda (a b) 
+        (let (
+            (res1 (permute a iv))
+            (res2 (permute (+ b (car res1)) (cdr res1)))
+            )
+            (car res2))))
+
+    ; hash a list of elements
+    ; such that (hashn '(a b)) = (hash2 a b)
+    (hashn (lambda (alist) 
+        (if (eq alist nil) nil
+        (let (
+            (res0 (permute (car alist) iv))
+            )
+            (letrec (
+                (iter (lambda (blist a b) 
+                    (if (eq blist nil) a
+                        (let (
+                            (res (permute (+ (car blist) a) b))
+                            )
+                            (iter (cdr blist) (car res) (cdr res))))))
+                )
+                (iter (cdr alist) (car res0) (cdr res0)))))))
+
+    ; basic 8-leaf tree
+    (tree8 (lambda (a b c d e f g h)
+        (hash2 
+            (hash2 (hash2 a b) (hash2 c d)) 
+            (hash2 (hash2 c d) (hash2 e f))))) 
+
+    ; proof of membership check
+    (treeproof8 (lambda (root x h1 h2 h3)
+        (eq root (hash2 h1 (hash2 h2 (hash2 h3 x))))))
```

Comment:
I like this a lot as an example. One nice thing about Lurk is that if you replace `hash2` with `cons`, you'll get real Poseidon hashing much more efficiently. But your proof will still make sense. In that case, you will be able to avoid hashing all the leaves because the destructuring/opening operators (`car` and `cdr`) have access to the non-deterministic information in the store — so they wouldn't have to be supplied. In that way, the prover's knowledge is encapsulated in the store.

Put differently, Lurk makes obvious that merkle proofs are exactly membership tests. In this case, the 'container' is an 8-leaf binary tree. So if we were to hard-code the leaf in question like your example, we could just write:

```
(lambda (root x)
  (eq x (cdr (cdr (cdr root)))))
```

Of course that requires that `root` be a `cons`. However, you can still use a cons (an opaque one) as a structured commitment. Just discard the type tag and use the value.

Since we have primitives for commitments and get some useful properties from them, you could also (at the cost of one extra hash to get the explicit hiding).

Then you would have
```
> (let ((tree8 (lambda (a b c d e f g h)
                 (hide 12345 (cons (cons (cons a b)
                                         (cons c d))
                                   (cons (cons e f)
                                         (cons g h))))))) 
  (tree8 1 2 3 4 5 6 7 8))
[100 iterations] => (comm 0x14e93dde5d0dbbdcccbebda5e91708f13d3e1fd846dc7b3a4ddf14d0ae1e012d)
> (let ((treeproof8 (lambda (root x)
                      (eq x (cdr (cdr (cdr (open root))))))))
    (treeproof8 0x14e93dde5d0dbbdcccbebda5e91708f13d3e1fd846dc7b3a4ddf14d0ae1e012d 8))
[24 iterations] => T
```

`open` in `treeproof8` works for the same reason `car` and `cdr` does; the prover _must_ know the opening and also the secret (in this case `12345`). In the case of an explicit commitment however, the apparent value returned to the user is evidently opaque, without extra error-prone manipulation of the cons.

---

I know these examples don't replace what you were trying to do here. I just wrote them out in case they help you get a better picture of why some Lurk features are as they are, and what they enable.